File: bindings/wasm/hierarchies_wasm/examples/src/validation/01_get_accreditations.ts
import { Federation, FederationProperty, PropertyName, PropertyValue } from "@iota/hierarchies/node";
import { strict as assert } from "assert";
import { generateRandomAddress, getFundedClient } from "../util";
export async function getAccreditations(): Promise<void> {
    const hierarchies = await getFundedClient();
    const { output: federation }: { output: Federation } = await hierarchies.createNewFederation().buildAndExecute(
        hierarchies,
    );

    console.log("\n✅ Federation created successfully!");
    console.log("Federation ID: ", federation.id);

    const propertyName = new PropertyName(["Example LTD"]);
    const propertyValue = PropertyValue.newText("Hello");

    await hierarchies.addProperty(
        federation.id,
        new FederationProperty(propertyName).withAllowedValues([propertyValue]),
    )
        .buildAndExecute(hierarchies);
    console.log(`\n✅ Property ${propertyName.dotted()} added successfully`);

    const receiver = generateRandomAddress();

    // Create and get the accreditation to attest
    const propertyToAttest = new FederationProperty(propertyName).withAllowedValues([PropertyValue.newText("Hello")]);

    await hierarchies.createAccreditationToAttest(federation.id, receiver, [propertyToAttest])
        .buildAndExecute(hierarchies);
    console.log(`\n✅ Accreditation to attest created for ${receiver}`);

    const accreditationsToAttest = await hierarchies.readOnly().getAccreditationsToAttest(federation.id, receiver);
    assert(accreditationsToAttest.accreditations.length > 0, "No accreditations to attest found");
    assert(
        accreditationsToAttest.accreditations[0].properties.length > 0,
        "No properties found in accreditation to attest",
    );
    assert.equal(
        accreditationsToAttest.accreditations[0].properties[0].propertyName.dotted(),
        propertyName.dotted(),
        "Property name does not match for accreditation to attest",
    );

    console.log("✅ Successfully retrieved accreditations to attest for the receiver:", receiver);

    // Create and get the accreditation to accredit

    const propertyToAccredit = new FederationProperty(propertyName).withAllowedValues([PropertyValue.newText("Hello")]);

    await hierarchies.createAccreditationToAccredit(federation.id, receiver, [propertyToAccredit])
        .buildAndExecute(hierarchies);
    console.log(`\n✅ Accreditation to accredit created for ${receiver}`);

    const accreditationsToAccredit = await hierarchies.readOnly().getAccreditationsToAccredit(federation.id, receiver);
    assert(accreditationsToAccredit.accreditations.length > 0, "No accreditations to accredit found");
    assert(
        accreditationsToAccredit.accreditations[0].properties.length > 0,
        "No properties found in accreditation to accredit",
    );
    assert.equal(
        accreditationsToAccredit.accreditations[0].properties[0].propertyName.dotted(),
        propertyName.dotted(),
        "Property name does not match for accreditation to accredit",
    );

    console.log("✅ Successfully retrieved accreditations to accredit for the receiver:", receiver);
}

File: hierarchies-rs/examples/validation/01_get_accreditations.rs
// Copyright 2020-2025 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use std::collections::HashSet;

use anyhow::Context;
use hierarchies::core::types::property::FederationProperty;
use hierarchies::core::types::property_name::PropertyName;
use hierarchies::core::types::property_value::PropertyValue;
use hierarchies::core::types::timespan::Timespan;
use hierarchies_examples::get_funded_client;
use iota_sdk::types::base_types::ObjectID;
use product_common::core_client::CoreClient;

/// Demonstrates how to use the offchain API to get accreditations to attest and accredit.
///
/// In this example we connect to a locally running private network, but it can be adapted
/// to run on any IOTA node by setting the network and faucet endpoints.
///
/// See the following instructions on running your own private network
/// https://github.com/iotaledger/iota/blob/develop/docs/content/developer/getting-started/connect.md

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let hierarchies_client = get_funded_client().await?;

    let federation = hierarchies_client
        .create_new_federation()
        .build_and_execute(&hierarchies_client)
        .await?;
    let federation_id = federation.output.id.object_id();

    let user_id = hierarchies_client.sender_address().into();

    let accreditations = hierarchies_client
        .get_accreditations_to_attest(*federation_id, user_id)
        .await?;

    println!("Accreditations to attest: {accreditations:#?}");

    //  Add Property
    let property_name = PropertyName::from("Example LTD");
    let value = PropertyValue::Text("Hello".to_owned());
    let allowed_values = HashSet::from_iter([value]);

    hierarchies_client
        .add_property(*federation_id, property_name.clone(), allowed_values.clone(), false)
        .build_and_execute(&hierarchies_client)
        .await
        .context("Failed to add Property")?;

    // Add new receiver
    let receiver = ObjectID::random();

    // Properties
    let properties = FederationProperty {
        name: property_name.clone(),
        allowed_values,
        shape: None,
        allow_any: false,
        timespan: Timespan::default(),
    };

    // Let us issue a accreditation to attest to the Property
    {
        hierarchies_client
            .create_accreditation_to_attest(*federation_id, receiver, vec![properties.clone()])
            .build_and_execute(&hierarchies_client)
            .await
            .context("Failed to issue accreditation to attest")?;
    }

    // Check if the accreditation to attest was issued
    let accreditations = hierarchies_client
        .get_accreditations_to_attest(*federation_id, receiver)
        .await
        .context("Failed to find accreditation to attest")?;

    assert!(accreditations.accreditations.len() == 1);

    println!("Accreditations to attest: {accreditations:#?}");

    // Issue accreditation to accredit
    {
        hierarchies_client
            .create_accreditation_to_accredit(*federation_id, receiver, vec![properties])
            .build_and_execute(&hierarchies_client)
            .await
            .context("Failed to issue accreditation to accredit")?;
    }

    // Check if the accreditation to accredit was issued
    let accreditations = hierarchies_client
        .get_accreditations_to_accredit(*federation_id, receiver)
        .await
        .context("Failed to find accreditation to accredit")?;

    assert!(accreditations.accreditations.len() == 1);

    Ok(())
}
